README.md
djaodjin-saas is a Django application that implements the logic to support
subscription-based Software-as-a-Service businesses.

Major Features:

- Separate billing profiles and authenticated users
- Double entry book keeping ledger
- Flexible security framework

Tested with

- **Python:** 3.7, **Django:** 3.2 ([LTS](https://www.djangoproject.com/download/)), **Django Rest Framework:** 3.12
- **Python:** 3.10, **Django:** 4.0 (latest), **Django Rest Framework:** 3.12
- **Python:** 2.7, **Django:** 1.11 (legacy), **Django Rest Framework:** 3.9.4

This project contains bare bone templates which are compatible with Django
and Jinja2 template engines. To see djaodjin-saas in action as part
of a full-fledged subscription-based session proxy, take a look
at [djaoapp](https://github.com/djaodjin/djaoapp/).

Full documentation for the project is available at
[Read-the-Docs](http://djaodjin-saas.readthedocs.org/)


Development
===========

After cloning the repository, create a virtualenv environment and install
the prerequisites:

<pre><code>
    $ python -m venv .venv
    $ source .venv/bin/activate
    $ pip install -r testsite/requirements.txt

    # Installs Javascript prerequisites to run in the browser
    $ make vendor-assets-prerequisites

</code></pre>

To use the testsite, you will need to add the payment processor keys
(see [Processor Backends](http://djaodjin-saas.readthedocs.io/en/latest/backends.html))
and Django secret key into a credentials file. Example with
[Stripe](https://stripe.com/):

<pre><code>
    $ cat ./credentials

    SECRET_KEY = "<em>enough_random_data</em>"
    STRIPE_PUB_KEY = "<em>your_stripe_public_api_key</em>"
    STRIPE_PRIV_KEY = "<em>your_stripe_private_api_key</em>"

</code></pre>

It remains to create and [populate the database with required objects](https://djaodjin-saas.readthedocs.io/en/latest/getting-started.html#setting-up-a-software-as-a-service-site).

    $ python ./manage.py migrate --run-syncdb --noinput
    $ python ./manage.py loaddata testsite/fixtures/initial_data.json
    $ python ./manage.py createsuperuser

You can further generate a set of dummy data data to populate the site.

    $ python ./manage.py load_test_transactions

Side note: If create your own fixtures file (ex: testsite/fixtures/test_data.json)
and attempt to load them with a Django version *before* 2 while the Python
executable was linked with a SQLite version *after* 3.25, you might stumble upon
the well-known [SQLite 3.26 breaks database migration ForeignKey constraint, leaving <table_name>__old in db schema](http://djaodjin.com/blog/django-2-2-with-sqlite-3-on-centos-7.blog.html#sqlite-django-compatibility) bug.
Your best bet is to use Django2+ or delete the migrations/ directory.

If all is well then, you are ready to run the server and browse the testsite.

    $ python manage.py runserver

    # Browse http://localhost:8000/


Implementation Notes
--------------------

The latest versions of django-restframework (>=3.0) implement paginators
disconnected from parameters in  views (i.e. no more paginate_by). You will
thus need to define ``PAGE_SIZE`` in your settings.py

    $ diff testsite/settings.py
    +REST_FRAMEWORK = {
    +    'PAGE_SIZE': 25,
    +    'DEFAULT_PAGINATION_CLASS':
    +        'rest_framework.pagination.PageNumberPagination',
    +}

This Django App does not send notification e-mails itself. All major
updates that would result in a e-mail sent trigger signals though. It is
straightforward to send e-mails on a signal trigger in the main
Django project. We provide sample e-mail templates here in the
saas/templates/notification/ directory.


Release Notes
=============

0.12.0

  * adds abstract models for `Organization` and `Role`
  * maps foreign keys to `settings.ORGANIZATION_MODEL` instead of directly mapping them to `saas.Organization`
  * uses `get_organization_model()` and `get_role_model()` instead of directly using `Organization` and `Role`
  * adds missing migrations from previous version

[previous release notes](changelog)


changelog
Roadmap
=======

1.0
  * fix group buy issues (balance statement, #155, #154) - done in 0.11.0
  * move views under saas.urls.views and update 'getting-started.rst'.
  * adds <profile> in charge API
  * rename <organization> into <profile> in url paths
  * support for Vue3
  * adds APIs for UseCharge

Released versions
=================

0.12.0

  * adds abstract models for `Organization` and `Role`
  * maps foreign keys to `settings.ORGANIZATION_MODEL` instead of directly mapping them to `saas.Organization`
  * uses `get_organization_model()` and `get_role_model()` instead of directly using `Organization` and `Role`
  * adds missing migrations from previous version

 -- Khalid Muhammad <khalid.refaat98@gmail.com>  Tue, 19 Jul 2022 18:00:00

0.11.0

  * fixes balance when extending a subscription through a group buy
  * fixes line items when checkout a group buy (#154, #155)
  * adds engaged/unengaged subscribers APIs
  * moves create shadow profile to /api/accounts/profiles/
  * adds inital compatibility with Django4.0

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Tue, 24 May 2022 19:47:00 -0700

0.10.3

  * adds API to retrieve inactive profiles
  * fixes multiple issues with create profile and optional fields

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Tue,  7 Dec 2021 03:12:00 -0700

0.10.2

  * restores backward compatibility - subscriptions key in profile detail API

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Thu, 27 Oct 2021 12:46:00 -0700

0.10.1

  * moves serializers.UserSerializer to serializers_overrides.UserSerializer
  * enables broker profile manager to update global RoleDescriptions
  * normalizes phone number before saving them into the database
  * prevents use of Coupon above nb_attempts limit (fix)
  * enables to customize html/css for pagination links

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Thu, 30 Sep 2021 00:00:00 -0700

0.10.0

  * works with with Django3.2
  * prepares compatibility with Vue3
  * moves all i18n translation server-side
  * renames `subscriptions` key into `subscribes_to` as it points to a Plan
  * categorizes a double-POST of the same Stripe token as a CardError

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Sun, 18 Jul 2021 17:30:00 -0700

0.9.5

  * keeps user model fields in sync with personal billing profile
  * prints last4 as exactly 4 digits on receipts
  * fixes index error in the presence of 100% discount and advance payments
  * fixes statement balances in the presence of GroupBuy
  * fixes PaymentIntent not found while using StripeConnect accounts
  * redirects to login page on redirects-to-profile if user not authenticated

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Mon, 26 Apr 2021 22:25:00 -0700

0.9.4

  * adds processor in checkout and card APIs
  * fixes computed balance due when subscription is free
  * re-raises all StripeError as ProcessorError

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Thu, 25 Mar 2021 11:01:00 -0700

0.9.3

  * migrates to Stripe PaymentIntent for SCA
  * adds balance due in checkout page
  * computes customers lifetime value (LTV) for a provider
  * reports error details when the payment processor has issues

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Tue, 23 Feb 2021 09:30:00 -0700

0.9.2

  * fixes translation of Transaction description when using period discounts

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Wed,  2 Sep 2020 10:43:00 -0700

0.9.1

  * translates Transaction description on display
  * fixes regression when creating code through the API
  * increases length of processor keys to 255

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Tue,  1 Sep 2020 17:34:00 -0700

0.9.0

  * reworks discount with an AdvanceDiscount model
  * adds pricing api
  * compatibles with Django3
  * fixes recording CartItem twice

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Thu,  7 Jun 2020 15:30:00 -0700

0.8.4

  * claims organization with no roles based on e-mail address
  * fixes implicit role grants

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Thu, 5 Mar 2020 09:35:00 -0700

0.8.3

  * adds implicit role to user when none exists
  * uses class name instead of HTML node indexes as hooks in checkout pages
  * uses consistent balance_amount/balance_unit accross APIs
  * replaces django-urldecorators by djaodjin-rules in testsite

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Wed, 22 Jan 2020 10:15:00 -0700

0.8.2

  * fixes add roles and subscriptions in multiple profile/user configurations
  * works around for Django/SQLite bug (#230)

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Fri, 13 Dec 2019 13:25:00 -0700

0.8.0

  * uses inline-template for Vue.js components
  * uploads profile picture
  * signals when processor is not connected and a charge is attempted

 -- Sebastien Mirolo <smirolo@djaodjin.com>  Sat, 26 Oct 2019 13:10:00 -0700

0.7.0
  * adds marketplace processing fee
  * restricts coupon to plan or number of use through UI
  * fixes invite new organizations with erroneous slugs

 -- Sebastien Mirolo <smirolo@djaodjin.com>   Sat, 14 Sep 2019 02:18:56 -0700

0.6.3
  * updates card-on-file directly in /billing/cart/ page
  * makes /api/accounts/ a search API instead of a typeahead API
  * fixes default search fields when looking for subscribers
  * adds remove card-on-file button in page templates

 -- Sebastien Mirolo <smirolo@djaodjin.com>   Tue, 13 Aug 2019 15:12:00 -0700

0.6.2
  * enables multiple StripeConnect authorize url (live/test)
  * fixes infinitescroll
  * fixes connect DELETE method to unsubscribes

 -- Sebastien Mirolo <smirolo@djaodjin.com>   Wed, 7 Aug 2019 00:00:00 -0700

0.6.1
  * improves API documentation
  * creates billing profile implicitely in various workflows
  * adds hooks for infinitescroll

 -- Sebastien Mirolo <smirolo@djaodjin.com>   Tue, 30 Jul 2019 00:00:00 -0700

0.6.0
  * supports sort by multiple columns (#182)
  * adds day_periods function (#199)
  * displays invited/requested roles first

saas/admin.py
# Copyright (c) 2020, DjaoDjin inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from django.contrib import admin

from .models import (AdvanceDiscount, Agreement, CartItem, Charge, ChargeItem,
    Coupon, RoleDescription, Plan, Signature, Subscription, Transaction)
from .utils import get_organization_model, get_role_model

admin.site.register(AdvanceDiscount)
admin.site.register(Agreement)
admin.site.register(CartItem)
admin.site.register(Charge)
admin.site.register(ChargeItem)
admin.site.register(Coupon)
admin.site.register(get_organization_model())
admin.site.register(Plan)
admin.site.register(get_role_model())
admin.site.register(RoleDescription)
admin.site.register(Signature)
admin.site.register(Subscription)
admin.site.register(Transaction)


saas/management/commands/compile_stats.py
# Copyright (c) 2016, DjaoDjin inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""Command for the cron job. Daily statistics"""

import datetime

from django.core.management.base import BaseCommand
from django.contrib.auth import get_user_model

from ...utils import get_organization_model


class Command(BaseCommand):
    """Daily usage for the service"""
    help = 'Print daily usage'

    def handle(self, *args, **options):
        end_period = datetime.datetime.now()
        start_period = end_period - datetime.timedelta(days=30)
        self.stdout.write('from %s to %s\n' % (start_period, end_period))
        for user in get_user_model().objects.filter(
            date_joined__gt=start_period):
            self.stdout.write('%s %s %s\n' % (str(user.date_joined),
                user.username, user.email))

        self.stdout.write('\n')
        for organization in get_organization_model().objects.filter(
            created_at__gt=start_period):
            self.stdout.write('%s %s\n'
                % (organization.created_at, organization))


saas/management/commands/delete_organization.py
# Copyright (c) 2017, DjaoDjin inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""
Command to obliterate all traces of an Organization,
including Transaction history.
"""

from django.core.management.base import BaseCommand

# We need this import to avoid getting an exception importing 'saas.models'
from ...utils import datetime_or_now, get_organization_model  # pylint: disable=unused-import


class Command(BaseCommand):

    help = "Obliterate all traces of an Organization."

    def handle(self, *args, **options):
        #pylint: disable=too-many-locals
        organizations = get_organization_model().objects.filter(slug__in=args)
        organizations.delete()


saas/migrations/0015_v0_11_0.py
# Generated by Django 3.2.14 on 2022-07-19 14:17

from django.db import migrations, models
import django.db.models.deletion
import django_countries.fields


class Migration(migrations.Migration):

    dependencies = [
        ('saas', '0014_v0_9_3'),
    ]

    operations = [
        migrations.AlterField(
            model_name='balanceline',
            name='selector',
            field=models.CharField(blank=True, help_text='Filter on transaction accounts', max_length=255),
        ),
        migrations.AlterField(
            model_name='cartitem',
            name='email',
            field=models.CharField(blank=True, help_text='e-mail of the person that will benefit from the subscription (GroupBuy)', max_length=255, null=True),
        ),
        migrations.AlterField(
            model_name='cartitem',
            name='option',
            field=models.PositiveIntegerField(default=0, help_text='Index in the list of discounts for advance payments'),
        ),
        migrations.AlterField(
            model_name='cartitem',
            name='plan',
            field=models.ForeignKey(help_text='Item in the cart (if plan)', null=True, on_delete=django.db.models.deletion.CASCADE, to='saas.plan'),
        ),
        migrations.AlterField(
            model_name='cartitem',
            name='sync_on',
            field=models.CharField(blank=True, help_text='identifier of the person that will benefit from the subscription (GroupBuy)', max_length=255, null=True),
        ),
        migrations.AlterField(
            model_name='charge',
            name='description',
            field=models.TextField(help_text='Description for the charge as appears on billing statements', null=True),
        ),
        migrations.AlterField(
            model_name='charge',
            name='processor_key',
            field=models.SlugField(help_text='Unique identifier returned by the payment processor', max_length=255, unique=True),
        ),
        migrations.AlterField(
            model_name='charge',
            name='state',
            field=models.PositiveSmallIntegerField(choices=[(0, 'created'), (1, 'done'), (2, 'failed'), (3, 'disputed')], default=0, help_text='Current state (i.e. created, done, failed, disputed)'),
        ),
        migrations.AlterField(
            model_name='chargeitem',
            name='invoiced_processor_fee',
            field=models.ForeignKey(help_text='Fee transaction to processor in order to process the transaction invoiced through this charge', null=True, on_delete=django.db.models.deletion.PROTECT, related_name='invoiced_processor_fee_item', to='saas.transaction'),
        ),
        migrations.AlterField(
            model_name='coupon',
            name='description',
            field=models.TextField(blank=True, help_text='Free-form text description for the coupon', null=True),
        ),
        migrations.AlterField(
            model_name='coupon',
            name='discount_type',
            field=models.PositiveSmallIntegerField(choices=[(1, 'percentage'), (2, 'currency'), (3, 'period')], default=1),
        ),
        migrations.AlterField(
            model_name='coupon',
            name='discount_value',
            field=models.PositiveIntegerField(default=0, help_text='Amount of the discount'),
        ),
        migrations.AlterField(
            model_name='organization',
            name='country',
            field=django_countries.fields.CountryField(max_length=2, verbose_name='Country'),
        ),
        migrations.AlterField(
            model_name='organization',
            name='default_timezone',
            field=models.CharField(default='America/Chicago', help_text='Timezone to use when reporting metrics', max_length=100),
        ),
        migrations.AlterField(
            model_name='plan',
            name='broker_fee_percent',
            field=models.PositiveIntegerField(default=0, help_text='Broker fee per transaction (in per 10000).'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='created_at',
            field=models.DateTimeField(auto_now_add=True, help_text='Date/time of creation (in ISO format)'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='description',
            field=models.TextField(help_text='Description of the plan'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='discontinued_at',
            field=models.DateTimeField(blank=True, help_text='Date/time the plan was discountinued (in ISO format)', null=True),
        ),
        migrations.AlterField(
            model_name='plan',
            name='is_active',
            field=models.BooleanField(default=False, help_text='True when a profile can subscribe to the plan'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='optin_on_request',
            field=models.BooleanField(default=False, help_text='True requires a provider to accept the subscription when created by the subscriber'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='organization',
            field=models.ForeignKey(help_text='Profile the plan belongs to', on_delete=django.db.models.deletion.CASCADE, related_name='plans', to='saas.organization'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='period_length',
            field=models.PositiveSmallIntegerField(default=1, help_text='Number of periods for a subscription to the plan (defaults to 1)'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='period_type',
            field=models.PositiveSmallIntegerField(choices=[(1, 'HOURLY'), (2, 'DAILY'), (3, 'WEEKLY'), (4, 'MONTHLY'), (5, 'YEARLY')], default=5, help_text='Natural period length of a subscription to the plan (hourly, daily, weekly, monthly, yearly)'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='renewal_type',
            field=models.PositiveSmallIntegerField(choices=[(0, 'ONE-TIME'), (1, 'AUTO-RENEW'), (2, 'REPEAT')], default=1, help_text='What happens at the end of a subscription period (one-time, auto-renew, repeat)'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='skip_optin_on_grant',
            field=models.BooleanField(default=False, help_text='True requires a subscriber to accept the subscription when created by the provider'),
        ),
        migrations.AlterField(
            model_name='plan',
            name='slug',
            field=models.SlugField(help_text='Unique identifier shown in the URL bar', unique=True),
        ),
        migrations.AlterField(
            model_name='plan',
            name='title',
            field=models.CharField(help_text='Title of the plan', max_length=50, null=True),
        ),
        migrations.AlterField(
            model_name='plan',
            name='unit',
            field=models.CharField(default='usd', help_text='Three-letter ISO 4217 code for currency unit (ex: usd)', max_length=3),
        ),
        migrations.AlterField(
            model_name='role',
            name='created_at',
            field=models.DateTimeField(auto_now_add=True, help_text='Date/time of creation (in ISO format)'),
        ),
        migrations.AlterField(
            model_name='role',
            name='grant_key',
            field=models.SlugField(blank=True, help_text='Key to identify the grant of the role', max_length=40, null=True),
        ),
        migrations.AlterField(
            model_name='role',
            name='request_key',
            field=models.SlugField(blank=True, help_text='Key to identify the request for the role', max_length=40, null=True),
        ),
        migrations.AlterField(
            model_name='roledescription',
            name='skip_optin_on_grant',
            field=models.BooleanField(default=False, help_text='Automatically grants the role without requiring a user to accept it.'),
        ),
        migrations.AlterField(
            model_name='subscription',
            name='organization',
            field=models.ForeignKey(help_text='Profile subscribed to the plan', on_delete=django.db.models.deletion.CASCADE, related_name='subscriptions', to='saas.organization'),
        ),
        migrations.AlterField(
            model_name='transaction',
            name='dest_organization',
            field=models.ForeignKey(help_text='Billing profile to which funds are deposited', on_delete=django.db.models.deletion.PROTECT, related_name='incoming', to='saas.organization'),
        ),
        migrations.AlterField(
            model_name='transaction',
            name='orig_organization',
            field=models.ForeignKey(help_text='Billing profile from which funds are withdrawn', on_delete=django.db.models.deletion.PROTECT, related_name='outgoing', to='saas.organization'),
        ),
    ]
